File: 230. Kth Smallest Element in a BST/solution-iterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        visited = []
        stack = []
        node = root
        
        # Use inorder traversal
        while node or stack:
            # break if the kth element is found
            if len(visited) == k:
                break
            if node:
                # check left
                stack.append(node)
                node = node.left
            else:
                # check self
                node = stack.pop()
                visited.append(node.val)
                
                # check right
                node = node.right
        
        return visited[-1]

File: 230. Kth Smallest Element in a BST/solution-iterative.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode node = root;
        while (k > 0 && (node != null || !stack.isEmpty())) {
            if (node != null) {
                stack.add(node);
                node = node.left;
            } else {
                while (node == null) {
                    node = stack.pop();
                }
                
                k--;
                if (k == 0) {
                    return node.val;
                } else {
                    node = node.right;
                }
            }       
        }
        return -1;
    }
}
